action_server/src/robocorp/action_server/_server_expose.py

```diff
@@ -0,0 +1,113 @@
+import asyncio
+import logging
+import json
+import httpx
+
+import websockets
+from fastapi import FastAPI
+from fastapi.testclient import TestClient
+from pydantic import BaseModel
+from ._settings import get_settings
+
+log = logging.getLogger(__name__)
+
+
+class SessionPayload(BaseModel):
+    sessionId: str
+    sessionSecret: str
+
+
+class BodyPayload(BaseModel):
+    requestId: str
+    path: str
+    method: str = "GET"
+    body: dict | None = None
+
+
+def forward_request(client: TestClient, payload: BodyPayload) -> httpx.Response:
+    if payload.method == "GET":
+        return client.get(payload.path)
+    elif payload.method == "POST":
+        return client.post(payload.path, json=payload.body)
+    elif payload.method == "PUT":
+        return client.put(payload.path, json=payload.body)
+    elif payload.method == "DELETE":
+        return client.delete(payload.path)
+    else:
+        raise NotImplementedError(f"Method {payload.method} not implemented")
+
+
+async def expose_server(app: FastAPI):
```

Comment:
Hummm... hadn't we agreed that this server forwarding would be in its own process?

In this case you'd create a new file and call it with `subprocess.call([sys.executable, ...])`.


Note: it should pass its pid to the child so that the child can automatically exit when the parent process exits -- the code I have around to do that is below (you can add it to the `robocorp.action_server._robo_utils.process` module):

```
_track_pids_to_exit = set()
_watching_thread_global = None


def exit_when_pid_exists(pid):
    _track_pids_to_exit.add(pid)
    global _watching_thread_global
    if _watching_thread_global is None:
        import time

        def watch_parent_process():
            # exit when any of the ids we're tracking exit.
            while True:
                for pid in _track_pids_to_exit:
                    if not is_process_alive(pid):
                        # Note: just exit since the parent process already
                        # exited.
                        log.info(
                            f"Force-quit process: %s because parent: %s exited",
                            os.getpid(),
                            pid,
                        )
                        os._exit(0)

                time.sleep(PARENT_PROCESS_WATCH_INTERVAL)

        _watching_thread_global = threading.Thread(target=watch_parent_process, args=())
        _watching_thread_global.daemon = True
        _watching_thread_global.start()
```

Also, then you can use `requests` instead of `httpx` in that case so that the new dep isn't needed.

Reply:
ok, refactoring.

But what are we gaining with separate process? Why not just run it in the same process / event loop?

Reply:
Also, `subprocess.call` is a blocking call, so using `subprocess.Popen` instead

Reply:
>But what are we gaining with separate process? Why not just run it in the same process / event loop?

The benefit is working with actual concurrency.

Reply:
I agree maybe the benefit is not that big (but I also think that the change would be small...).

Reply:
This is done. please review!

Reply:
This part seems good. Resolving conversation.